### ISSUE TITLE
Peer refused to agree to our IP address

### ISSUE THREAD

**Issue by DimitriPapadopoulos** [adrienverge/openfortivpn#920]
I wasn't able to connect this morning. Part of the log:
```
INFO:   Negotiation complete.
Peer refused to agree to our IP address
Connect time 0.4 minutes.
Sent 1225 bytes, received 1219 bytes.
Connection terminated.
Modem hangup
ERROR:  read: Input/output error
INFO:   Cancelling threads...
INFO:   Cleanup, joining threads...
ERROR:  pppd: The link was terminated by the modem hanging up.
INFO:   Terminated pppd.
INFO:   Closed connection to gateway.
INFO:   Logged out.
```

Looks similar to #362.

**Comment by DimitriPapadopoulos** [comment]
Works now. Perhaps a temporary malfunction of the VPN server?

In any case, the error originates in `pppd` and then propagates to _openfortivpn_.

**Comment by maloute** [comment]
I have the same issue, it's random but I need to login 2 to 10 times to not have this error and to connect to my VPN. I don't currently have access to my pc but i'm using openfortivpn with an ubuntu distribution. Will post the precise version number tonight.
Is there any steps to do precise troubleshooting on this issue ?

**Comment by DimitriPapadopoulos** [comment]
I don't really have a workaround. As far as I can see, the error originates in `pppd`.

**Comment by sigbertklinke** [comment]
Hi, I have the same problem under Ubuntu 20.04.4. Sometimes it works, sometimes not :( The FortiClient works.  

Sigbert 

```
INFO:   Connected to gateway.
INFO:   Authenticated.
INFO:   Remote gateway has allocated a VPN.
Using interface ppp0
Connect: ppp0 <--> /dev/pts/1
INFO:   Got addresses: [172.18.203.131], ns [141.20.1.3, 141.20.2.3]
INFO:   negotiation complete
INFO:   negotiation complete
Peer refused to agree to our IP address
Connect time 0.1 minutes.
Sent 1175 bytes, received 1169 bytes.
Connection terminated.
Modem hangup
```

**Comment by DimitriPapadopoulos** [comment]
If @skrynauw is correct, I'm afraid the only solution would be to get rid of `pppd` and instead use embedded PPP code and a `tun` device.

**Comment by sigbertklinke** [comment]
Just an additional remark: In two subsequent calls it worked once and then not :(

```
Using interface ppp0
Connect: ppp0 <--> /dev/pts/1
INFO:   Got addresses: [172.18.203.3], ns [141.20.1.3, 141.20.2.3]
INFO:   negotiation complete
INFO:   negotiation complete
local  IP address 172.18.203.3
remote IP address 192.0.2.1
INFO:   Interface ppp0 is UP.
INFO:   Setting new routes...
INFO:   Adding VPN nameservers...
INFO:   Tunnel is up and running.
```
```
INFO:   Remote gateway has allocated a VPN.
Using interface ppp0
Connect: ppp0 <--> /dev/pts/1
INFO:   Got addresses: [172.18.203.3], ns [141.20.1.3, 141.20.2.3]
INFO:   negotiation complete
INFO:   negotiation complete
Peer refused to agree to our IP address
Connect time 0.1 minutes.
Sent 1175 bytes, received 1169 bytes.
Connection terminated.
Modem hangup

```

**Comment by randomstuff** [comment]
This is happening randomly for me as well. This does not appear to be a conflict of IP address: my LAN is using 192.0.0.0/24 while the VPN network assigns 172.18.203.103.

**Comment by DimitriPapadopoulos** [comment]
The error occurs inside `pppd`:
https://github.com/ppp-project/ppp/commit/7360193b12811b3497b79cb1ca74a9c89203bc8e

It might be worth investigating whet happens inside `pppd`, by studying or modifying the `pppd` source code or attaching a debugger to the `pppd` process.

Perhaps additional parameters passed to `pppd` might be enough to fix the error:
https://github.com/adrienverge/openfortivpn/blob/ace8696d4a105639acab71884fb64785ac24c81f/src/tunnel.c#L236-L250

Finally the version of FortiOS on the gateway might be of interest, as this might be a bug in the gateway itself, that for some reason shows up with `pppd` but not with the PPP implementation in the official client. For that reason, please [always provide](https://github.com/adrienverge/openfortivpn/wiki#reporting-issues):
- version of operating system,
- version of openfortivpn,
- version of FortiOS on the VPN gateway (retrieve it with `openfortivpn -v -v -v`),
- full logs (redacted if needed) of `openfortivpn` (`-v -v -v`) and of `pppd` (`--pppd-log`).

**Comment by IvanLloro** [comment]
Indeed, adding "ipcp-accept-local" in "/etc/ppp/options" solves the issue for me and executing "openfortivpn -c /etc/openfortivpn/config" works.

In my case, I was receiving:
INFO:   Remote gateway has allocated a VPN.
Using interface ppp0
Connect: ppp0 <--> /dev/pts/1
INFO:   Got addresses: [10.1.1.6], ns [10.X.X.X, 10.X.X.X]
INFO:   negotiation complete
INFO:   negotiation complete
INFO:   negotiation complete
INFO:   negotiation complete
INFO:   negotiation complete
Peer refused to agree to our IP address
[...]
Connection terminated.
Modem hangup

My system is a Linux Gentoo and the relevant pppd package is  net-dialup/ppp version  2.4.9-r8.

Finally, just to simplify Google searchability of this fix, I am not using the CLI for setting a VPN but the KDE NetworkManager applet (which ends up using networkmanager-fortisslvpn plugin) to connect to a Fortinet server with 2-factor-authentication (2FA). For this setup to work, it is also needed to ensure the 2FA checkbox within the Advanced configuration button in KDE configuration is left unchecked (or just delete the relevant line in /etc/NetworkManager/system-connections/[VPN_connection_configuration_file] in order to be able to connect (if not, NetworkManager will fail even before calling openfortivpn).

**Comment by DimitriPapadopoulos** [comment]
Interesting. I could add **`ipcp-accept-local`** here:
https://github.com/adrienverge/openfortivpn/blob/ace8696d4a105639acab71884fb64785ac24c81f/src/tunnel.c#L236-L250

I am not 100 % certain it won't break anything, but let's test it.

That said, I don't understand why the "peer would refuse to agree to our local address":
```
# With this option, pppd will accept the peer's idea of our local IP
# address, even if the local IP address was specified in an option.
#ipcp-accept-local
```
Indeed we do not specify a local address:
https://github.com/adrienverge/openfortivpn/blob/ace8696d4a105639acab71884fb64785ac24c81f/src/tunnel.c#L239
Instead, we explicitly ask the peer to supply the local IP address:
```
# Disables the default behaviour when no local IP address is specified,
# which is to determine (if possible) the local IP address from the
# hostname. With this option, the peer will have to supply the local IP
# address during IPCP negotiation (unless it specified explicitly on the
# command line or in an options file).
#noipdefault
```

**Comment by DimitriPapadopoulos** [comment]
It might be that file **`/etc/ppp/options`** has been modified on your computer and does define an IP local address.

**Comment by Altap826-pro** [comment]
Randomstuff has went up very price. Should get rid of all PPPS
